File: 116.Populating_Next_Right_Pointers_in_Each_Node.java
/*
 // Definition for a Node.
 class Node {
 public int val;
 public Node left;
 public Node right;
 public Node next;
 
 public Node() {}
 
 public Node(int _val,Node _left,Node _right,Node _next) {
 val = _val;
 left = _left;
 right = _right;
 next = _next;
 }
 };
 */
class Solution {
    public Node connect(Node root) {
        Node head = root;   //for return
        Node curr = new Node();
        while (root != null && root.left != null) {
            curr = root;
            while (curr != null) {
                curr.left.next = curr.right;
                if (curr.next == null) curr.right.next = null;
                else curr.right.next = curr.next.left;
                curr = curr.next;
            }
            root = root.left;
        }
        return head;
    }
}

File: 116.PopulatingNextRightPointersinEachNode.py
class Solution:
    def connect(self, root: 'Node') -> 'Node':
        if not root:
            return None
        queue = [root]
        while queue:
            n = len(queue)
            for i in range(n):
                node = queue.pop(0)
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
                if i < n - 1:
                    node.next = queue[0]
        return root